File: sdk-js/src/keys.ts
/**
 * Key management (blueprint §11, §5 verb card).
 *
 *   generateKey(opts)   → POST /api/v1/keys/generate
 *   getKey(keyId)       → GET  /api/v1/keys/{id}
 *   listKeys()          → GET  /api/v1/keys
 *   rotateKey(keyId)    → POST /api/v1/keys/rotate/{id}
 *   revokeKey(keyId)    → POST /api/v1/keys/revoke/{id}
 *
 * GATEWAY GAP: the gateway's rotate/revoke routes take no request body, so the
 * Python SDK's `grace_period` / `reason` arguments cannot be forwarded over REST
 * yet. They are accepted here for forward-compatibility but not sent.
 */

import { getConfig } from "./config.js";
import { request } from "./http.js";
import {
  toKeyInfo,
  type KeyInfo,
  type WireKeyGenerateRequest,
  type WireKeyInfo,
  type WireKeyListResponse,
  type WireKeyResponse,
} from "./types.js";

export interface GenerateKeyOptions {
  algorithm?: string;
  /** Key purpose: "encryption" (default) or "signing". */
  type?: string;
  organization?: string;
  owner?: string;
  metadata?: Record<string, string>;
}

export interface RotateKeyOptions {
  /** Overlap window during which the old key still decrypts (gateway gap: unused). */
  gracePeriod?: string;
}

export interface RevokeKeyOptions {
  /** Reason recorded with the revocation (gateway gap: unused). */
  reason?: string;
}

export async function generateKey(opts: GenerateKeyOptions = {}): Promise<KeyInfo> {
  const config = getConfig();
  const reqBody: WireKeyGenerateRequest = {
    algorithm: opts.algorithm ?? config.defaultAlgorithm ?? null,
    type: opts.type ?? "encryption",
    organization: opts.organization ?? "",
    owner: opts.owner ?? "",
    metadata: opts.metadata ?? {},
  };
  const resp = await request<WireKeyResponse>("/api/v1/keys/generate", {
    method: "POST",
    body: reqBody,
  });
  return toKeyInfo(resp);
}

export async function getKey(keyId: string): Promise<KeyInfo> {
  const resp = await request<WireKeyResponse>(`/api/v1/keys/${encodeURIComponent(keyId)}`);
  return toKeyInfo(resp);
}

export async function listKeys(): Promise<KeyInfo[]> {
  const resp = await request<WireKeyListResponse>("/api/v1/keys");
  return (resp.keys ?? []).map((k: WireKeyInfo) => toKeyInfo(k));
}

export async function rotateKey(
  keyId: string,
  _opts: RotateKeyOptions = {},
): Promise<Record<string, unknown>> {
  return request<Record<string, unknown>>(`/api/v1/keys/rotate/${encodeURIComponent(keyId)}`, {
    method: "POST",
  });
}

export async function revokeKey(
  keyId: string,
  _opts: RevokeKeyOptions = {},
): Promise<Record<string, unknown>> {
  return request<Record<string, unknown>>(`/api/v1/keys/revoke/${encodeURIComponent(keyId)}`, {
    method: "POST",
  });
}

File: sdk-python/privyq/keys.py
"""Key management wrappers (BP §11.2, §16)."""

from __future__ import annotations

from typing import Any, Mapping

from ._proto import privyq_pb2 as pb
from .client import Client, get_default_client
from .config import get_config


def generate_key(
    *,
    algorithm: str | None = None,
    key_type: str = "encryption",
    organization: str = "",
    owner: str = "",
    metadata: Mapping[str, Any] | None = None,
    client: Client | None = None,
):
    client = client or get_default_client()
    req = pb.GenerateKeyRequest(
        algorithm=algorithm or get_config().default_algorithm,
        type=key_type,
        organization=organization,
        owner=owner,
        metadata={str(k): str(v) for k, v in (metadata or {}).items()},
    )
    return client.call("GenerateKey", req).key


def get_key(key_id: str, *, client: Client | None = None):
    """Fetch public key info by id (the v2 ``get_key`` verb; core ``GetPublicKey``)."""
    client = client or get_default_client()
    return client.call("GetPublicKey", pb.GetPublicKeyRequest(key_id=key_id)).key


def rotate_key(key_id: str, *, grace_period: str = "24h", client: Client | None = None):
    client = client or get_default_client()
    return client.call("RotateKey", pb.RotateKeyRequest(key_id=key_id, grace_period=grace_period))


def revoke_key(key_id: str, *, reason: str = "", client: Client | None = None):
    client = client or get_default_client()
    return client.call("RevokeKey", pb.RevokeKeyRequest(key_id=key_id, reason=reason))
